### DISCUSSION TITLE
using tokei as a library without filesystem access?

### DISCUSSION THREAD

**Discussion by ahornby** [XAMPPRocky/tokei#744] [Q&A]
Any pointers to example code or docs on how to passing a path and in memory content and get stats without tokei touching the filesystem (there are no files there, directories do not exist locally). 

I've found LanguageType::parse_from_slice() but wonder if there is a higher level entry point to consider using (or maybe adding?).

**Comment by XAMPPRocky** [comment]
Hello, Thank you for your issue! I've transferred this over to tokei's discussion board, as it feels more appropriate. 

As it relates to your question. I'm curious what you think would be a higher level abstraction than `LanguageType::parse_from_slice`?. It does no file system access and parses the given data as the language you set.

**Reply by ahornby** [reply]
I couldn't find an exist method to infer the LanguageType for an in memory buffer+pathname without filesystem access.  LanguageType::from_path("./main.rs", &Config::default()) can load the file from disk, but would need to read from a buffer instead.

Perhaps a new method that's something like  LanguageType::from_slice(path_thats_not_on_disk, &buffer_of_file_contents, &config)?

**Reply by XAMPPRocky** [reply]
Oh okay, you can get the language from the the extension of the path with [`Language::from_file_extension`](https://docs.rs/tokei/12.1.2/tokei/enum.LanguageType.html#method.from_file_extension)

**Reply by ahornby** [reply]
That would miss out on the languages which do shebang/content based detection though?

**Reply by ahornby** [reply]
Would you be interested in a PR to add an in memory form of LanguageType::from_path that does the shebang based/other content based detection in memory?    e.g. LanguageType::from_slice(path_thats_not_on_disk, &buffer_of_file_contents, &config)?

**Reply by XAMPPRocky** [reply]
I'm not sure about a from_slice, I think just a `from_shebang_path` would be enough and if people want both, they can use both. Tokei doesn't do any other kind of content based file detection.
